Fab (Floating Action Button) Interactor

elrickvm opened the pull request:
## Motivation
Floating action button is one of the staples from the Google Material UI Interface used in both Mobile and Desktop Applications. Users use this Fab to execute primary action within the top hierarchal context for a particular task or section they are in.

## Approach
- Create Fab Interactor
- Leverage the tests and examples for Button Interactor PR #4 to show that all the associated classes from the Material UI can be accessed and tested. So we don't need to do that in these test for the Floating Action button.

elrickvm committed:
Create Fab Interactor

elrickvm committed:
Export Fab Interactor

elrickvm committed:
Fix conflict

cowboyd commented:
I'm curious how we'll handle the locators for FABs that only have a `+` icon. Will we use the aria-label?

elrickvm committed:
Add Material Icons

elrickvm committed:
Update test

elrickvm commented:
That is good question. @cowboyd I added a few ways to try to get the Icon.

1. Added a filter to the Interactor that looks for the classname on the svg
2. Added a filter to get the aria-label
3. Created a Icon Interactor in the test and use that to assert for the presence of the Icon. ( presently that fails )
  --- This failure is the same thing that I am experiencing with Radio and Button around disabled. For some reason the Query is not returning anything. So the assertions fail. If I open the test the window and try the same QuerySelectorAll like bigTest it fails until i open the inspector and click on the element and retry the same Query and it passes. 
  
Do you have an idea of what could be causing the issue of DOM elements being present but the QuerySelector fails until you click on the element?

elrickvm committed:
Add more ways to assert for Icon

wKich commented:
@elrickvm, how about using the `.MuiFab-root` selector for locator?

elrickvm commented:
@wKich I think using `.MuiFab-root` in this case?

elrickvm commented:
@wKich do you mean using `.MuiFab-root` as the selector for the Interactor like so? Not sure how to use the class as a locator cause I think the locator needs to return a string right @cowboyd ?

```js
export default Button.extend('button')
  .selector('button.MuiFab-root')
  .filters({
    ariaLabel: (el) => el.getAttribute("aria-label"),
    svgIcon: (el) => el.querySelector('svg')?.classList.contains('MuiSvgIcon-root')
  })

wKich commented:
@elrickvm yeah, use it as a selector. Sorry for misword )

elrickvm commented:
No problem @wKich I mix the two sometimes as well that is why I had to ask...:-).

elrickvm committed:
Use mui selector to grab fab

cowboyd commented on test/fab.test.tsx:
```diff
@@ -0,0 +1,47 @@
+    .assertion(fab.has({ ariaLabel: including('add')}))
+    .assertion(fab.has({ svgIcon: true}))
+    // TODO: Fix this assertion fails with:: "ERROR actual.includes is not a function"
+    // .assertion(fab.find(SvgIcon({ className: including('MuiSvgIcon-root') })).exists())
```
Do SVG elements necessarily have a class name? or is it in some weird namespace like `svg:className` 🤔

elrickvm replied:
@cowboyd I think so. It returns something different when you try to access the className in the DOM. I moved the check into a filter.

elrickvm committed:
Update test

elrickvm committed:
Remove export

wKich reviewed (commented):
Looks good to me 👍

jnicklas commented on src/fab.ts:
```diff
@@ -0,0 +1,8 @@
+  .selector('button.MuiFab-root')
+  .filters({
+    ariaLabel: (el) => el.getAttribute("aria-label"),
+    svgIcon: (el) => el.querySelector('svg')?.classList.contains('MuiSvgIcon-root')
```
It might be better to cast this to a boolean. If the SVG element is not found this would return `undefined`.

```suggestion
    svgIcon: (el) => !!el.querySelector('svg')?.classList.contains('MuiSvgIcon-root')
```
